### PR TITLE
Fix bug in rematerialization safety logic

### PR DESCRIPTION
**Pull request by andrewcraik**
The RematTools server to allow you to try and recompute a store on a different control flow path. This is used in the inliner to do privatized inliner argument rematerialization in the blocks on the taken side of virtual guards. The RematTools include a bunch of safety checking logic to ensure it is safe to remat a store. There is a bug in this remat logic that could cause some nodes that should be checked to be skipped. This patch fixes this problem and adds additional helpful trace to the RematTools.

### CODE REVIEW

File: compiler/optimizer/RematTools.cpp [hunk 1]
```diff
@@ -190,21 +190,35 @@ void RematTools::walkNodesCalculatingRematSafety(TR::Compilation *comp,
...
 
       walkNodesCalculatingRematSafety(comp, currentNode->getChild(i),
             scanTargets, enabledSymRefs, unsafeSymRefs, trace, visitedNodes);
       }
 
+   if (trace)
+      {
+      traceMsg(comp, "Remat Safety Walk visiting n%dn\n", currentNode->getGlobalIndex());
+      traceMsg(comp, "Enabled symrefs: ");
+      (*comp) << enabledSymRefs;
```
Comment (fjeremic) [inline compiler/optimizer/RematTools.cpp]: Cool! Did not know compilation had overloaded the stream operator.

**Comment by mgaudet** [comment]
I think you have a typo in your commit message ("The RematTools serve**r**") :smile:

**Comment by andrewcraik** [comment]
Fixed

**Comment by vijaysun-omr** [comment]
Looks like a clear bug fix; we were iterating over the children of the node but checking if the node was visited every time rather than the children.

**Pull request merged**